Feature/ra ladder rankings

CO2-code opened the pull request:
Added the feature of displaying top 3 RA rankers on 1v1 and 2v2 ladders with their UI, Code needs cleaning , edits and reviews 
also make it only if clientgametype=RA...

CO2-code committed:
Update LANLobby.cs

CO2-code committed:
Update LANLobby.cs

CO2-code committed:
Update LANLobby.cs

CO2-code committed:
Update LANLobby.cs

CO2-code committed:
Update LANLobby.cs

CO2-code committed:
Update LANLobby.cs

CO2-code committed:
Update LANLobby.cs

CO2-code committed:
Update LANLobby.cs

CO2-code committed:
ladder

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Fix syntax error in JoinUser method

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Refactor FetchAndDisplayLaddersAsync method

Refactor ladder fetching and display logic to improve error handling and UI updates.

CO2-code committed:
Fix error handling in CnCNetLobby

CO2-code committed:
Fix formatting and add missing newline at end of file

CO2-code committed:
Fix indentation and formatting in CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Fix formatting and add missing closing brace

CO2-code committed:
Fix exception handling in CnCNetLobby

CO2-code committed:
Fix exception handling in CnCNetLobby

CO2-code committed:
Refactor ladder fetching to update labels directly

CO2-code committed:
Remove unnecessary closing brace in CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Fix exception handling in CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Refactor ladder fetching and UI update logic

Refactor ladder fetching logic to simplify and improve error handling. Update UI labels directly with fetched data.

CO2-code committed:
Refactor FetchAndDisplayLaddersAsync method

CO2-code committed:
Refactor ladder fetching and display logic

CO2-code committed:
Fetch and display ladder rankings with updates

CO2-code committed:
Refactor ladder fetching and display logic

CO2-code committed:
Add using directive for Microsoft.Xna.Framework

CO2-code committed:
Refactor spriteBatch initialization in CnCNetLobby

Refactored spriteBatch initialization to use a using statement for better resource management.

CO2-code committed:
Optimize Draw method for ladder background rendering

Refactor Draw method to initialize ladder background texture only once and improve error handling.

CO2-code committed:
Introduce ladderBgAlpha for dynamic background opacity

Added ladderBgAlpha variable to control background transparency.

CO2-code committed:
Update ladder background color to white

CO2-code committed:
Refactor ladder background color setting

CO2-code committed:
Change ladder background color to dark gray

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Refactor ladder background variables and trophy names

CO2-code committed:
Use Clamp method for ladderBgAlpha calculation

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

11EJDE11 commented:
Appreciate the effort, CO2. Bespoke things like a ladder UI would be better done through INI files IMO. To help with this I think we should create an XNAClientJSONLabel which could be set like:

```
[lblLadderRankings]
$Type=XNAClientJSONLabel
Template=Ladder players: {YR[0].player_name} ({YR[0].points:N0}), {YR[1].player_name} ({YR[1].points:N0}), {YR[2].player_name} ({YR[2].points:N0})
LoadingText=Ladder players: Loading...
URL=https://ladder.cncnet.org/api/v1/qm/ladder/rankings
RefreshIntervalSeconds=500
TimeoutSeconds=10
FallbackText=Ladder unavailable
FontIndex=1
AnchorPoint=716,148
```

Then it's reusable - you could use it to display news or updates, ladder players, etc and it's all configured through the ini. I can put one together if there's support for it? @SadPencil @Metadorius @MahBoiDeveloper

CO2-code commented:
> Appreciate the effort, CO2. Bespoke things like a ladder UI would be better done through INI files IMO. To help with this I think we should create an XNAClientJSONLabel which could be set like:
> 
> ```
> [lblLadderRankings]
> $Type=XNAClientJSONLabel
> Template=Ladder players: {YR[0].player_name} ({YR[0].points:N0}), {YR[1].player_name} ({YR[1].points:N0}), {YR[2].player_name} ({YR[2].points:N0})
> LoadingText=Ladder players: Loading...
> URL=https://ladder.cncnet.org/api/v1/qm/ladder/rankings
> RefreshIntervalSeconds=500
> TimeoutSeconds=10
> FallbackText=Ladder unavailable
> FontIndex=1
> AnchorPoint=716,148
> ```
> 
> Then it's reusable - you could use it to display news or updates, ladder players, etc and it's all configured through the ini. I can put one together if there's support for it? @SadPencil @Metadorius @MahBoiDeveloper 

The things I added can be easily configured from LANLobby.ini just like the others (lbRA1v1title , lbRA1v1 (same for 2v2) ,and lbLadderRankings )

CO2-code commented:
I didn't upload LANLobby.ini mb

CO2-code committed:
Update LANLobby.ini

CO2-code committed:
Merge branch 'CnCNet:develop' into feature/RA-ladder-rankings

CO2-code committed:
Update CnCNetLobby.ini

Changing the chat msgs size leaves a space at the top in Lan lobby , so I separated them :)

CO2-code committed:
Update CnCNetLobby.ini

CO2-code committed:
Update LANLobby.ini

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Merge branch 'CnCNet:develop' into feature/RA-ladder-rankings

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs

CO2-code committed:
Update CnCNetLobby.cs
